arch/x86/kvm/pkvm/pgtable.c

```diff
@@ -362,3 +465,45 @@ int pkvm_pgtable_map(struct pkvm_pgtable *pgt, unsigned long vaddr,
 	data.phys = ALIGN_DOWN(phys, PAGE_SIZE);
 	return pkvm_pgtable_walk(pgt, vaddr, size, &walker);
 }
+
+/**
+ * pkvm_pgtable_unmap() - Remove a virtual addr to physcal addr mapping from a
+ *			  pkvm pgtable.
+ * @pgt:	The page table to remove mapping.
+ * @vaddr:	The virtual address of the mapping to be removed.
+ * @phys:	The desired physical address for the unmap walker to verify.
+ * @size:	The memory size to unmap.
+ *
+ * The unmapped range is the minimum PAGE_SIZE-aligned range covering
+ * [@vaddr, @vaddr + @size), and @phys is aligned down to the PAGE_SIZE if @phys
+ * is not INVALID_PAGE.
+ *
+ * All the mappings for the memory pages in [@vaddr, @vaddr + @size) are removed
+ * from the page table. Each leaf entry in the range is visited by the unmap
+ * walker, to unmap the present ones, followed by the TLB flushing. If the range
+ * represented by a present leaf overlaps with the unmapped range, the leaf will
+ * be split into smaller ones for the unmap walker to walk down to the next
+ * level to remove. If all the mappings in a page table page are removed then
+ * this page table page will be freed after the TLB flushing.
+ *
+ * When @phys is not INVALID_PAGE, the unmap walker will verify if the physical
+ * address in the removed matches with @phys or not. If @phys is INVALID_PAGE
```

Comment:
s/in the removed/in the removed mapping/ ?

Also it doesn't explain what exactly "verify" means, i.e. doesn't say what happens if verification fails (i.e. doesn't say that the hypervisor will panic).

Also, actually it verifies not just that `@vaddr` maps to `@phys` but that the entire `[@vaddr, @vaddr + @size)` maps contiguously to `[@phys, @phys + @size)`, right? So it's better to clarify that?

Reply:
Make sense. I will add more words to explain and clarify "verify".